walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,83 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
 
 public final class FindMeetingQuery {
+  
+  Collection<TimeRange> timeOptions = new ArrayList<TimeRange>();
+  
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+    
+    if (request.getAttendees().isEmpty() || events.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+    
+    Set<String> allEventAttendees = new HashSet<>();
+    for (Event event : events) {
+      allEventAttendees.addAll(event.getAttendees());
+    }
+    Set<String> allRequestAttendees = new HashSet<>();
+    allRequestAttendees.addAll(request.getAttendees());
+    for (String attendee : allRequestAttendees) {
+      if (!allEventAttendees.contains(attendee)) {
+        return Arrays.asList(TimeRange.WHOLE_DAY);
+      }
+    }
+    
+    List<Event> eventsList = new ArrayList<Event>();
+    eventsList.addAll(events);
+    Collections.sort(eventsList, Event.ORDER_BY_START);
+    boolean earlyEnd = false;
+    if (TimeRange.START_OF_DAY != eventsList.get(0).getWhen().start()) {
+      addTimeRange(TimeRange.START_OF_DAY, eventsList.get(0).getWhen().start());
+    }
+    for (int i = 1; i < eventsList.size(); i++) {
+      if (eventsList.get(i - 1).getWhen().end() != eventsList.get(i).getWhen().start()) {
```

Comment:
Consider factoring the contents of this loop into a function accepting a pair of events.

(all these (i - 1), (i + 1) are hard for me to understand without a deep reading of the code)